### ISSUE TITLE
@children Only Working in Chrome

### ISSUE THREAD

**Issue by TonyLugg** [aurelia/templating#655]
Using `@children` in a custom element is only working in Chrome.  Adding `@useShadowDOM` to the custom element got it working in FireFox but it does not work in Edge.  My elements have slots and two types of custom elements can be nested within each other forming a multi-level menu.

@fkleuver This is a carry over from our discussion in the vNext space.

**Comment by fkleuver** [comment]
@TonyLugg Thanks. Can you please post your markup and the piece of your typescript code that declares and consumes the `@children` decorator?

**Comment by fkleuver** [comment]
I did some digging and it seems the implementation differs based on whether you're having `@useShadowDOM` or not, but that implementation does not account for whether the browser actually supports shadowDOM or not (which Edge doesn't).

So for starters, you should probably either remove that `@useShadowDOM` decorator (apologies for the wrong advice there) or, if you do actually need to use shadow dom, polyfill it in Edge.

**Comment by fkleuver** [comment]
@bigopon you probably want to know about this too. Previous discussion at https://github.com/aurelia/aurelia/issues/331

**Comment by TonyLugg** [comment]
```
<menu-list>
	<div>This is a menu</div>
	<menu-list-item>List item number 1</menu-list-item>
	<menu-list-item>List item number 2</menu-list-item>
	<menu-list-item>Sub menu 3
		<menu-list>
			<menu-list-item>Sub item number 1</menu-list-item>
			<menu-list-item>Sub item number 2</menu-list-item>
		</menu-list>
	</menu-list-item>
	<menu-list-item>List item number 4</menu-list-item>
</menu-list>
```

```
export class MenuListCustomElement {
	el: Element;
	@children('menu-list-item') items: any[];
	parentId: number;

	constructor(el: Element) {
		this.el = el;
		this.parentId = Math.floor((Math.random() * 1000000) + 1);
	}

	attached() {
		console.log('items', this.items);
		if (this.items) {
			for (let loItem of this.items) {
				console.log('loItem:', loItem);
				if (loItem.el && loItem.el.au && loItem.el.au.controller && loItem.el.au.controller.viewModel) {
					let vm = <MenuListItemCustomElement>loItem.el.au.controller.viewModel;
					console.log('loItem vm:', vm);
					console.log('uniqueId', vm.uniqueId);
					console.log('submenu', vm.subMenu);
				}
			}
		}
	}
}	
```

**Comment by TonyLugg** [comment]
In Chrome I actually don't need to reference el.au.controller.viewModel, as the vm's are right there in the array of children.

**Comment by StrahilKazlachev** [comment]
1. So the issue is with items that are initially present? Not ones added later on?
2. Also the html markup is how you are *using* it?
2.1 `<menu-list>` has a `<slot>`?

**Comment by TonyLugg** [comment]
@StrahilKazlachev Correct.  All items in View HMTL.  Each custom element has a slot (menu-list and menu-list-item).  menu-list can create multiple menu-list-item's, menu-list-item can contain visual content and a single menu-item (a sub menu).

**Comment by fkleuver** [comment]
Yeah if you use slots then the view compiler enables ShadowDOM mode for the custom elements. 
@StrahilKazlachev you are on point then about what you said on Discord.

@TonyLugg You could either try polyfilling ShadowDOM in Edge with shadydom, or you can use `replaceable` instead of slots to ensure ShadowDOM is not being used.

That said, this is still a valid issue as I believe we should be able to detect whether something is *really* shadowDOM or not, and if not, use the appropriate initialization logic.

**Comment by EisenbergEffect** [comment]
For vCurrent, you can use slots without ShadowDOM. If you do, the framework attempts to emulate. If you specify `@useShdadowDOM` and your browser doesn't support it, it also tries to emulate it. But, perhaps specifying `@useShdadowDOM` when there's no native support causes the issue with `@children`. I'd either polyfill ShadowDOM or remove `@useShadowDOM` as quick fixes. Let me know if either of those solves the problem.

**Comment by TonyLugg** [comment]
@EisenbergEffect  I had originally not used `@useShadowDOM` and only Chrome was working for my use case.  I added `@useShadowDOM` and FF worked but still Edge did not work.

**Comment by EisenbergEffect** [comment]
Can you create a simple repro based on the Aurelia CLI and attach it here? I'd like to take a look myself. This seems pretty surprising to me so either there's something not quite right in your project or we've had a major regression at some point.

**Comment by TonyLugg** [comment]
OK, I'm tied up for a few days but will get something for you soon.

**Comment by TonyLugg** [comment]
I just created a little sample in Aurelia Typescript Sandbox and it works [there](https://codesandbox.io/s/325jplnzlp).  My real test app is using CLI Webpack config.  Also, the custom elements are global resources.

**Comment by TonyLugg** [comment]
I created a new CLI app with Webpack and TypeScript and used the same files as the sandbox linked above.  Now I get the original results, i.e. it works in Chrome but not in FF or Edge.

**Comment by EisenbergEffect** [comment]
Do you mind zipping that up and attaching it here?

**Comment by TonyLugg** [comment]
[menu-test4Rob.zip](https://github.com/aurelia/templating/files/2704400/menu-test4Rob.zip)

Here is the test project, sans node modules.

**Comment by davismj** [comment]
https://gist.run/?id=040775f06aba5e955afd362ee60863aa

**Comment by davismj** [comment]
Firefox

![image](https://user-images.githubusercontent.com/3845823/50370509-661ada00-05eb-11e9-96ed-1f9adc732ae1.png)

Chrome

![image](https://user-images.githubusercontent.com/3845823/50370514-759a2300-05eb-11e9-850a-9f764180bfd2.png)

Edge 

![image](https://user-images.githubusercontent.com/3845823/50370525-b85bfb00-05eb-11e9-9f9a-af477d4ab4cc.png)

@TonyLugg The outputs in the console look identical to me. Can you help me understand what is different in Edge or Firefox, please?

**Comment by TonyLugg** [comment]
Correct.  Per the sandbox sample I provided it is working there too.  However, it does not work with the CLI Webpack config that I attached for Rob. [above](https://github.com/aurelia/templating/issues/655#issuecomment-449509949)

**Comment by TonyLugg** [comment]
I finished the multi-level menu using `querySelectorAll` and referencing `el.au.controller.viewModel` instead of `@children`.  I did not need to build a tree as each menu-list and menu-list-item referenced their children and parent.

**Comment by bigopon** [comment]
@TonyLugg Did you get your issue resolved?

**Comment by TonyLugg** [comment]
@bigopon The issue with `@child` and `@children` was not resolved to my knowledge.  I worked around it as mentioned above based on comments provided by @davismj.

**Comment by bigopon** [comment]
@TonyLugg I'm finally able to have a look at the cause of this. In short, this is a timing issue. It can be explained like this:
- **With native shadow on**: every element is kept as is, what you see is what you get initially. And with this, the child observer is able to initialize the value of `@child`/`@children` during `bind` lifecycle, hence you get what you want in `attached`.
- **With native shadow off**: content elements of a custom element will be removed during compilation, and added later during bind phase. This is to do the projection for `<slot/>` emulation, and it will trigger mutation observer and thus give you the value you want. The downside of this, is every mutation callback will be a tick later. And in your example, it's 3 layers of tick, while `attached` is only 1 tick after `bind`, hence undefined in attached as you saw. This, unfortunately is the only reliable way to have slot emulation work, so change handler should be encouraged in emulation scenario.
